File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque, defaultdict

def solution(n, edge):
    visited = [False for _ in range(n + 1)]
    adj_vertices = defaultdict(list)
    distance = [21427483647 for _ in range(n + 1)]
    distance[1] = 0
    max_distance = 0
    answer = 0
    
    for v1, v2 in edge:
        adj_vertices[v1].append(v2)
        adj_vertices[v2].append(v1)

    queue = deque()
    queue.append((1, 0))
    visited[1] = True
    
    while queue:
        v, d = queue.popleft()
        
        for adj_vertex in adj_vertices[v]:
            if (not visited[adj_vertex]):
                queue.append((adj_vertex, d + 1))
                distance[adj_vertex] = d + 1
                visited[adj_vertex] = True
    
    max_distance = max(distance[1:])
    answer = distance.count(max_distance)
    
    return answer

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.js
function solution(n, edge) {
    const visited = Array.from({ length: n + 1}, () => false);
    const adjVertices = Array.from({ length: n + 1 }, () => []);
    const distance = Array.from({ length: n + 1}, () => Number.MAX_SAFE_INTEGER)
    distance[1] = 0
    
    for (const [v1, v2] of edge) {
        adjVertices[v1].push(v2)
        adjVertices[v2].push(v1)
    }
    
    const queue = []
    queue.push([1, 0])
    visited[1] = true
    
    while (queue.length > 0) {
        const [v, d] = queue.shift();
        
        for (const adjVertex of adjVertices[v]) {
            if (!visited[adjVertex]) {
                queue.push([adjVertex, d + 1]);
                distance[adjVertex] = d + 1;
                visited[adjVertex] = true;
            }
        }
    }
    
    const maxDistance = Math.max(...distance.slice(1));
    const answer = distance.filter((dist) => dist === maxDistance).length;
    
    return answer;
    
}